syntax = "proto3";

package auction_house;

import "google/protobuf/timestamp.proto";

service AuctionHouse {
    rpc GetAuctions(Empty) returns (Auctions);
    rpc GetAuctionItems(GetAuctionItemsRequest) returns (AuctionItems);
    rpc SubscribeAuctions(SubscribeAuctionRequest) returns (stream SubscriptionUpdate);
    rpc UnsubscribeAuctions(SubscribeAuctionRequest) returns (stream SubscriptionStatus);
    rpc PlaceBid(PlaceBidRequest) returns (BidStatus);
}

message Empty {}

message Auction {
    uint32 id = 1;
    string title = 2;
    string description = 3;
    AuctionType type = 4;
    google.protobuf.Timestamp start_time = 5;
    google.protobuf.Timestamp end_time = 6;
    Location location = 7;

    message Location {
        string address = 1;
        string city = 2;
        bool online = 3;
    }
}

message Auctions {
    repeated Auction auctions = 1;
}

message AuctionItem {
    uint32 id = 1;
    uint32 auction_id = 2;
    string name = 3;
    string description = 4;
    float start_price = 5;
    float current_price = 6;
    bool sold = 7;
}

message AuctionItems {
    repeated AuctionItem items = 1;
}

message GetAuctionItemsRequest {
    uint32 auction_id = 1;
}

message SubscribeAuctionRequest {
    uint32 client_id = 1;
    repeated uint32 auction_ids = 2;
}

message PlaceBidRequest {
    uint32 client_id = 1;
    uint32 auction_id = 2;
    uint64 item_id = 3;
    float price = 4;
}

message SubscriptionStatus {
    bool success = 1;
    string message = 2;
}

message SubscriptionUpdate {
    int32 auctionId = 1;
    oneof update_type {
        AuctionItem newItem = 2;
        AuctionItem updatedItem = 3;
    }
}

message BidStatus {
    bool success = 1;
    float new_price = 2;
    string message = 3;
}

enum AuctionType {
    STANDARD = 0;
    PREMIUM = 1;
    WEEKLY = 2;
}
